File: backend/models/viewer.go
package models

import pb "goodanime-backend/proto"

// ToProto converts ViewerAvatar to its protobuf representation.
func (a *ViewerAvatar) ToProto() *pb.ViewerAvatar {
	if a == nil {
		return nil
	}
	return &pb.ViewerAvatar{
		Medium: a.Medium,
	}
}

// ToProto converts Viewer to its protobuf representation.
func (v *Viewer) ToProto() *pb.Viewer {
	if v == nil {
		return nil
	}
	return &pb.Viewer{
		Id:        v.ID,
		Name:      v.Name,
		CreatedAt: v.CreatedAt,
		Avatar:    v.Avatar.ToProto(),
	}
}

// ViewerAvatar represents the avatar for a viewer
type ViewerAvatar struct {
	Medium string `json:"medium"`
}

// Viewer represents a user on the platform
type Viewer struct {
	ID        int32        `json:"id"`
	Name      string       `json:"name"`
	CreatedAt int32        `json:"createdAt"`
	Avatar    ViewerAvatar `json:"avatar"`
}

// ViewerDTO represents the DTO for a viewer query
type ViewerDTO struct {
	Viewer Viewer `json:"Viewer"`
}

File: lib/models/viewer.dart
import '../proto/viewer.pb.dart' as pb;

/// Represents a user on the platform
class Viewer {
  /// User ID
  final int id;

  /// Display name
  final String name;

  /// Medium-sized avatar URL
  final String avatarMedium;

  /// Account creation timestamp
  final int createdAt;

  /// Creates a viewer
  const Viewer({
    required this.id,
    required this.name,
    required this.avatarMedium,
    required this.createdAt,
  });

  /// Creates a viewer from a protobuf object
  factory Viewer.fromProto(pb.Viewer pbObj) {
    return Viewer(
      id: pbObj.id,
      name: pbObj.name,
      avatarMedium: pbObj.hasAvatar() ? pbObj.avatar.medium : '',
      createdAt: pbObj.createdAt,
    );
  }

  /// Converts the viewer to a protobuf object
  pb.Viewer toProto() {
    final pbObj = pb.Viewer(id: id, name: name, createdAt: createdAt);
    if (avatarMedium.isNotEmpty) {
      final avatarPb = pb.ViewerAvatar(medium: avatarMedium);
      pbObj.avatar = avatarPb;
    }
    return pbObj;
  }
}
